core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+	block_import: Arc<Mutex<I>>,
+	client: Arc<C>,
+	mut env: E,
+	inherent_data_providers: inherents::InherentDataProviders,
+) where
+	I: BlockImport<B> + Send + Sync + 'static,
+	C: HeaderBackend<B> + AuxStore + ProvideRuntimeApi + 'static,
+	C::Api: PowApi<B>,
+	E: Environment<B> + Send + Sync + 'static,
+	E::Error: core::fmt::Debug,
+{
+	if !inherent_data_providers.has_provider(&srml_timestamp::INHERENT_IDENTIFIER) {
+		match inherent_data_providers.register_provider(srml_timestamp::InherentDataProvider) {
+			Ok(()) => (),
+			Err(e) => warn!("Registering inherent data provider for timestamp failed"),
+		}
+	}
+
+	thread::spawn(move || {
+		loop {
+			match mine_one(&block_import, &client, &mut env, &inherent_data_providers) {
+				Ok(()) => (),
+				Err(e) => warn!("Mining block failed: {:?}", e),
+			}
+
+			std::thread::sleep(std::time::Duration::new(1, 0));
+		}
+	});
+}
+
+pub fn mine_one<B: BlockT<Hash=H256>, I, C, E>(
+	block_import: &Arc<Mutex<I>>,
+	client: &Arc<C>,
+	env: &mut E,
+	inherent_data_providers: &inherents::InherentDataProviders,
+) -> Result<(), String> where
+	I: BlockImport<B>,
+	C: HeaderBackend<B> + AuxStore + ProvideRuntimeApi,
+	C::Api: PowApi<B>,
+	E: Environment<B>,
+	E::Error: core::fmt::Debug,
+{
+	let best_hash = client.info().best_hash;
+	let best_header = client.header(BlockId::Hash(best_hash))
+		.map_err(|e| format!("Best header does not exist: {:?}", e))?
+		.ok_or("Fetching best header failed")?;
+	let best_aux_key = POW_AUX_PREFIX.iter().chain(&best_hash[..])
+		.cloned().collect::<Vec<_>>();
+	let mut aux = match client.get_aux(&best_aux_key)
+		.map_err(|e| format!("{:?}", e))?
+	{
+		Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e))?,
+		None => Default::default(),
+	};
+	let mut proposer = env.init(&best_header).map_err(|e| format!("{:?}", e))?;
+
+	let inherent_data = inherent_data_providers
+		.create_inherent_data().map_err(String::from)?;
+	let block = futures::executor::block_on(proposer.propose(
+		inherent_data,
+		Default::default(),
+		std::time::Duration::new(1, 0)
+	)).map_err(|e| format!("Block proposing error: {:?}", e))?;
+
+	let (mut header, body) = block.deconstruct();
+	let (difficulty, seal) = client.runtime_api().mine(
+		&BlockId::Hash(best_hash),
+		&header.hash(),
+	).map_err(|e| format!("Mine sealing runtime error: {:?}", e))?;
+
+	aux.total_difficulty += difficulty;
+	let hash = header.hash();
+
+	let key = POW_AUX_PREFIX.iter().chain(&hash[..]).cloned().collect::<Vec<_>>();
+	let import_block = BlockImportParams {
+		origin: BlockOrigin::Own,
+		header,
+		justification: None,
+		post_digests: vec![DigestItem::Seal(POW_ENGINE_ID, seal)],
+		body: Some(body),
+		finalized: false,
+		auxiliary: vec![(key, Some(aux.encode()))],
+		fork_choice: ForkChoiceStrategy::Custom(true),
+	};
+
+	if let Err(err) = block_import.lock().expect("Lock failed").import_block(import_block, Default::default()) {
+		warn!("Error with block built on {:?}: {:?}",
```

Comment:
I am not sure, but shouldn't this be an `error!` level? The fact that we cannot import a block built by ourselves seems to me quite severe.

Reply:
Changed this to error out in `mine_loop`. The current behavior is that if the loop fails, it prints a message, wait one second before mining again.